### PR TITLE
Add support for per-profile tab colors

### LINKED ISSUE #1337: settings.json should provide tab color as part of a scheme and/or profile, with an optional way to introduce variant for multiple instances of the same profile
# Summary of the new feature/enhancement

Profiles.json should provide tab color as part of a scheme and/or profile, with an optional way to introduce variant for multiple instances of the same profile. When a user opens different tabs today, there is very little distinction between the tabs. I'd like to have the ability to apply a color to the tab, perhaps with an alpha so that I could shade both light and dark applications with a accent color representing the type of profile used by the tab. It might be valuable to have an additional accent color which can shade the tab or a portion of the tab uniquely depending on if there are multiple instances of the same profile open, so as to distinguish between them.

# Proposed technical implementation details

I'm not entirely sure how the optional portion would be implemented, so I'll address the first portion first. 

```JSON
{
  "profiles" :
  {
    "tabForeground" : "#aarrbbgg",
    "tabBackground" : "#aarrbbgg"
  }
}
```

or

```JSON
{
  "profiles" :
  {
    "colorScheme" : "Foo"
  },
  "schemes" :
  [
    {
      "name" : "Foo",
      "tabForeground" : "#aarrbbgg",
      "tabBackground" : "#aarrbbgg"
    }
  ]
}
```

Just as with the foreground and background properties, the profile definition overrides the scheme definition. This concept could be expanded later to use other foreground and background identifiers such as images.

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request
profile-tabColor-000

This PR adds support for per-profile tab colors, in accordance with #7134. This adds a single `tabColor` property, that when set, specifies the background color for profile's tab. This color can be overridden by the color picker, and clearing the color with the color picker will revert to this default color set for the tab.

## References
* Seriously go read #7134
* Theming is covered in #3327 & #5772 

## PR Checklist
* [x] Closes #1337
* [x] I work here
* [ ] Tests added/passed
* [ ] Requires documentation to be updated

## Detailed Description of the Pull Request / Additional comments

* Seriously go read #7134

## Validation Steps Performed
Played with setting this color, both on launch and via hot-reload

### CODE REVIEW

File: src/cascadia/TerminalApp/Profile.cpp [hunk 1]
```diff
@@ -231,6 +232,13 @@ TerminalSettings Profile::CreateTerminalSettings(const std::unordered_map<std::w
 
     terminalSettings.AntialiasingMode(_antialiasingMode);
 
+    if (_tabColor)
+    {
+        uint32_t c = _tabColor.value();
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/Profile.cpp:237]: ew, I should clean this back up

File: src/cascadia/TerminalApp/Tab.cpp [hunk 2]
```diff
@@ -707,114 +721,164 @@ namespace winrt::TerminalApp::implementation
...
     std::optional<winrt::Windows::UI::Color> Tab::GetTabColor()
     {
-        return _tabColor;
+        const auto currControlColor{ GetActiveTerminalControl().TabColor() };
+        std::optional<winrt::Windows::UI::Color> controlTabColor;
+        if (currControlColor != nullptr)
+        {
+            controlTabColor = currControlColor.Value();
+        }
+        return Utils::CoalesceOptionalsOrNot(_runtimeTabColor,
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/Tab.cpp:730]: add a comment better explaining this

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 3]
```diff
@@ -1818,16 +1818,16 @@ namespace winrt::TerminalApp::implementation
...
-                std::optional<Windows::UI::Color> color = tab->GetTabColor();
-                if (color.has_value())
-                {
-                    _SetNonClientAreaColors(color.value());
-                }
-                else
-                {
-                    _ClearNonClientAreaColors();
-                }
+                // // Raise an event that our titlebar color changed
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1821]: Oh no, dead code!

File: src/cascadia/TerminalApp/Profile.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Profile.cpp:237]: apparently, I can't just construct this inline, but w/e

File: src/cascadia/TerminalApp/Tab.cpp [hunk 4]
```diff
@@ -707,114 +721,179 @@ namespace winrt::TerminalApp::implementation
...
+            controlTabColor = currControlColor.Value();
+        }
+
+        // A Tab's color will be the result of layering a variety of sources,
+        // from the bottom up:
+        //
+        // Color                |             | Set by
+        // -------------------- | --          | --
+        // Runtime Color        | _optional_  | Color Picker / `setTabColor` action
+        // Control Tab Color    | _optional_  | Profile's `tabColor`, or a color set by VT
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Tab.cpp:737]: Oh so I'm thinking that this operates more like the tab title, which can be set by VT, but the VT title can also be overridden temporarily at runtime the tab renamer.

[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Tab.cpp:737]: Also note that I'm not sure that there _is_ a VT sequence for setting the tab color, that's purely a hypothetical

File: src/cascadia/TerminalApp/Tab.cpp [hunk 5]
```diff
@@ -8,13 +8,15 @@
 #include "Tab.g.cpp"
 #include "Utils.h"
 #include "ColorHelper.h"
+#include "../../types/inc/utils.hpp"
```
Comment (miniksa) [inline src/cascadia/TerminalApp/Tab.cpp:11]: Can we not? I think we're trying to put new stuff into TIL. I'm also not a big fan of including the `::Microsoft::Console` namespace in the Terminal App. We're trying to break from that.

File: src/types/inc/utils.hpp [hunk 6]
```diff
@@ -124,4 +124,28 @@ namespace Microsoft::Console::Utils
...
+    template<typename T>
+    std::optional<T> CoalesceOptionalsOrNot(const std::nullopt_t& base)
+    {
+        return base;
+    }
+
+    // Method Description:
+    // - Returns the value from the first populated optional, or the last one (if none of the previous had a value)
+    template<typename T, typename... Ts>
+    std::optional<T> CoalesceOptionalsOrNot(const std::optional<T>& t1, Ts&&... t2)
```
Comment (miniksa) [inline src/types/inc/utils.hpp:147]: Per other comment, consider moving to til. Add a unit test.

File: src/cascadia/TerminalApp/Tab.cpp (hunk 5 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Tab.cpp:11]: Yea I just stuck it where I found the original `CoalesceOptionals`. I can move it to `til` in this PR, or make another for specifically that, whichever you prefer @miniksa

File: src/types/inc/utils.hpp (hunk 6 above)
Comment (zadjii-msft) [reply src/types/inc/utils.hpp:147]: done

File: src/inc/til/coalesce.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+// Licensed under the MIT license.
+
+#pragma once
+
+namespace til
+{
+    // Method Description:
+    // - Base case provided to handle the last argument to CoalesceOptionals<T...>()
+    template<typename T>
+    T CoalesceOptionals(const T& base)
```
Comment (miniksa) [inline src/inc/til/coalesce.h:11]: The naming of these doesn't really fit the naming theme of other `til` things. Primarily the all lowercase with underscores if necessary, concise/short as possible names, and reusing terminology/monikers from STL where possible.

Maybe `til::coalesce` for `til::CoalesceOptionals` much like `std::optional<T>::value()` with the partner
`til::coalesce_or_not` or `til::coalesce_or_empty` much like `std::optional<T>::value_or(T)`

Comment (DHowett) [reply src/inc/til/coalesce.h:11]: maybe `til::coalesce(opt, opt, opt)` and `til::coalesced_value(opt, opt, opt, val)`. you can even rebuild coalesced_value on top of coalesce!

**Comment by zadjii-msft** [comment]
@DHowett and I discussed this and he's more okay with it now that we've considered the fact that the color needs to be associated with each individual control, not just per-tab. It's more like and extension of titlebar state:
* titlebar _text_
* titlebar _color_
* titlebar _icon_

etc.

File: src/cascadia/TerminalCore/ICoreSettings.idl [hunk 8]
```diff
@@ -34,6 +34,8 @@ namespace Microsoft.Terminal.TerminalControl
         String WordDelimiters;
 
         Boolean ForceVTInput;
+
+        Windows.Foundation.IReference<UInt32> TabColor;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalCore/ICoreSettings.idl:38]: Shouldn't this be an IControlSetting instead of an ICoreSetting?

Comment (zadjii-msft) [reply src/cascadia/TerminalCore/ICoreSettings.idl:38]: So, I was imagining this as just another thing that's similar to the title. If VT can modify the "title color" of the terminal, then that's something that's owned by the Core, not the control. It's the same way title "text" works now.

File: src/cascadia/TerminalApp/Tab.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jantari) [reply src/cascadia/TerminalApp/Tab.cpp:734]: > Also note that I'm not sure that there _is_ a VT sequence for setting the tab color, that's purely a hypothetical

One could argue setting the title to a text string with a VT-sequence to set the background color could reasonably be interpreted as a request / intent to set the tab (or titlebar if tabs are disabled )color.

**Comment by xorinzor** [comment]
@zadjii-msft could you please clarify how we can actually add this to our json settings per-profile now?
I simply tried adding `tabColor` to a profile, but that didn't do anything. The documentation isn't very clear in how to configure this either, and the UI doesn't have an option either.

Other issues I found all refer to this issue regarding per-profile tab colors, but haven't been able to find the answer.

**Pull request merged**